### PR TITLE
Adding application, api and enduser information to the warn log in authentication handler

### PR DESCRIPTION
**Pull request by chamikasudusinghe**
Fix: The complete solution can be provided in a very specific scenario where the JWT tokens are replaced with the default tokens. When we try to invoke an API without scope assignment, this warning is triggered and we can provide the requested logs. In other scenarios, when the warning is triggered, since we are using JWT, we do not get the application details since the token invalidation happens before getting the application name.

Solution: By getting content available in the messageContext to be displayed as part of the warn log.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java [hunk 1]
```diff
@@ -429,6 +429,13 @@ public boolean handleRequest(MessageContext messageContext) {
...
             }
+
+            // Get application information to be logged in API authentication failure (only requestURI for JWT tokens)
+            String applicationName = (String) messageContext.getProperty(APIMgtGatewayConstants.APPLICATION_NAME);
+            String endUserName = (String) messageContext.getProperty(APIMgtGatewayConstants.END_USER_NAME);
+            String requestURI = (String) messageContext.getProperty(RESTConstants.REST_FULL_REQUEST_PATH);
+            String warnDetails = "";
+
             if (log.isDebugEnabled()) {
                     // We do the calculations only if the debug logs are enabled. Otherwise this would be an overhead
```
Comment (dushaniw) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java:440]: shall we move the construction of warn log message to a separate method?

Comment (chamikasudusinghe) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java:434]: added a separate method and included the sample error strings

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java [hunk 2]
```diff
@@ -429,6 +429,7 @@ public boolean handleRequest(MessageContext messageContext) {
             } else if (Util.tracingEnabled()) {
                 Util.setTag(keyTracingSpan, APIMgtGatewayConstants.ERROR, APIMgtGatewayConstants.KEY_SPAN_ERROR);
             }
+
```
Comment (dushaniw) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java:432]: remove the additional line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java [hunk 3]
```diff
@@ -445,8 +446,14 @@ public boolean handleRequest(MessageContext messageContext) {
                     log.error("API authentication failure due to "
                             + APISecurityConstants.API_AUTH_GENERAL_ERROR_MESSAGE, e);
                 } else {
+                    String warnDetails = getWarnDetails(messageContext);
                     // We do not need to log known authentication failures as errors since these are not product errors.
-                    log.warn("API authentication failure due to " + errorMessage);
+                    log.warn("API authentication failure due to " + errorMessage + warnDetails);
+                    // Sample Error String (JWT): APIAuthenticationHandler API authentication failure due to The access
```
Comment (dushaniw) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java:452]:
```suggestion
                    // Sample Warn String (JWT): APIAuthenticationHandler API authentication failure due to The access
```
move the comment before the warn log

**Review by dushaniw — APPROVED** [review]

**Pull request merged**